WWDC/ImageDownloadCenter.swift

```diff
@@ -8,68 +8,130 @@
+/// Singleton coordinator for concurrent image downloads with caching and request deduplication via Operation dependencies.
+///
+/// Critical to the implementation is that an ImageRetrievalOperation is responsible for both cache retrieval and downloading (if necessary).
+///
+/// ```
+/// Request(URL)
+///      │
+///      ▼
+/// ┌─────────────┐
+/// │ Active Op?  │──Yes──→ Chain as Dependency ──→ Reuse Result
+/// └─────────────┘
+///      │ No
+///      ▼
+/// ┌─────────────┐
+/// │ Memory Cache│──Hit───→ Return Cached
+/// └─────────────┘
+///      │ Miss
+///      ▼
+/// ┌─────────────┐
+/// │ Disk Cache  │──Hit───→ Load + Store in Memory ──→ Return
+/// └─────────────┘
+///      │ Miss
+///      ▼
+/// ┌─────────────┐
+/// │   Network   │──────→ Download + Store All Caches ──→ Return
+/// └─────────────┘
+/// ```
+final class ImageDownloadCenter: Logging, Sendable {
 
     static let shared: ImageDownloadCenter = ImageDownloadCenter()
 
-    let cache = ImageCacheProvider()
+    private let cache = ImageCacheProvider()
 
     static let log = makeLogger(subsystem: ImageDownload.subsystemName, category: "ImageDownloadCenter")
 
-    private let dispatchQueue = DispatchQueue(label: "ImageDownloadCenter", qos: .userInitiated, attributes: .concurrent)
-    private lazy var queue: OperationQueue = {
-        let q = OperationQueue()
-        
-        q.underlyingQueue = dispatchQueue
+    private let dispatchQueue: DispatchQueue
+    private let queue: LockIsolated<OperationQueue>
+
+    init() {
+        let dispatchQueue = DispatchQueue(label: "ImageDownloadCenter", qos: .userInteractive, attributes: .concurrent)
+        let queue: OperationQueue = OperationQueue()
+        queue.underlyingQueue = dispatchQueue
+
+        self.dispatchQueue = dispatchQueue
+        self.queue = LockIsolated(queue)
+    }
 
-        return q
-    }()
-    
     func cachedThumbnail(from url: URL) -> NSImage? { cache.cachedImage(for: url, thumbnailOnly: true).thumbnail }
 
-    /// The completion handler is always called on the main thread
+    /// The completion handler is always called on the main thread if it's downloaded, but it's called on the callers thread if it comes from the cache
+    /// Cache retrieval happens synchronously, so the completion handler is called immediately if the image is cached. But that means you're hitting
+    /// the file system on the main thread.
     @discardableResult
-    func downloadImage(from url: URL, thumbnailHeight: CGFloat, thumbnailOnly: Bool = false, completion: @escaping ImageDownloadCompletionBlock) -> Operation? {
-        if thumbnailOnly {
-            if let thumbnailImage = cache.cachedImage(for: url, thumbnailOnly: true).thumbnail {
-                completion(url, (nil, thumbnailImage))
-                return nil
+    func downloadImage(from url: URL, thumbnailHeight: CGFloat, thumbnailOnly: Bool = false, completion: @escaping ImageDownloadCompletionBlock) -> Operation {
```

Comment:
Re-structuring this to always return an Operation is motivated by filesystem accesses for cached images blocking the main thread. This has been causing the very choppy scrolling of the sessions list.